src/main/java/com/linecorp/armeria/server/metrics/ServiceMetricDecorator.java

```diff
@@ -0,0 +1,151 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.metrics;
+
+import java.util.function.Supplier;
+
+import com.linecorp.armeria.common.ServiceInvocationContext;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.server.DecoratingServiceCodec;
+import com.linecorp.armeria.server.RequestTimeoutException;
+import com.linecorp.armeria.server.ServiceCodec;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.util.AttributeKey;
+import io.netty.util.concurrent.Promise;
+
+/**
+ * Decorator to collect service metrics.
+ *
+ * This class is expected to be used with other ServiceMetricConsumer
+ */
+public class ServiceMetricDecorator extends DecoratingServiceCodec {
+    private static final AttributeKey<Metrics> METRICS =
+            AttributeKey.valueOf(ServiceMetricDecorator.class, "METRICS");
+
+    private final ServiceMetricConsumer metricConsumer;
+
+    /**
+     * Creates a new instance that decorates the specified {@link ServiceCodec}.
+     */
+    public ServiceMetricDecorator(ServiceCodec codec, ServiceMetricConsumer consumer) {
+        super(codec);
+        metricConsumer = consumer;
+    }
+
+    @Override
+    public DecodeResult decodeRequest(Channel ch, SessionProtocol sessionProtocol, String hostname, String path,
+                                      String mappedPath, ByteBuf in, Object originalRequest,
+                                      Promise<Object> promise) throws Exception {
+        final long startTime = System.nanoTime();
+        final int requestSize = in.readableBytes();
+
+        DecodeResult decodeResult = delegate().decodeRequest(
+                ch, sessionProtocol, hostname, path, mappedPath, in, originalRequest, promise);
+
+        Supplier<Long> lazyElapsedTime = () -> System.nanoTime() - startTime;
+
+        switch (decodeResult.type()) {
+        case SUCCESS: {
+            ServiceInvocationContext context = decodeResult.invocationContext();
+            context.attr(METRICS).set(new Metrics(sessionProtocol, requestSize, startTime));
+
+            promise.addListener(future -> {
+                if (!future.isSuccess()) {
+                    // encodeFailureResponse will process this case.
+                    return;
+                }
+                Object result = future.getNow();
+
+                if (result instanceof FullHttpResponse) {
+                    FullHttpResponse httpResponse = (FullHttpResponse) result;
+                    metricConsumer.onComplete(
+                            Status.of(sessionProtocol, httpResponse.status().code()), lazyElapsedTime.get(),
```

Comment:
Could we just call `System.nanoTime()` directly to avoid potential unnecessary instantiation of a `Long` object?